packages/mui-material/test/integration/MenuList.test.js

```diff
@@ -498,7 +498,10 @@ describe('<MenuList> integration', () => {
         fireEvent.mouseDown(menuitem);
         menuitem.focus();
       });
-      fireEvent.click(menuitem);
+
+      await act(async () => {
+        fireEvent.click(menuitem);
+      });
```

Comment:
The tests are now passing except in jsdom, where 77 tests are failing, and it looks like most or all of them are the same case:

```
Expected test not to call console.error() but instead received 1 calls.

If the warning is expected, test for it explicitly by using the toErrorDev() matcher.
Test location:
  /home/romgrk/src/material-ui/packages/mui-material/test/integration/MenuList.test.js 

console.error message #1:
  Warning: An update to ForwardRef(TouchRipple) inside a test was not wrapped in act(...).
```

Because the ripple is lazy, it now runs inside a promise (microtask) handler:

```typescript
class LazyRipple {
  start(...args) {
    this.mount().then(() => this.ref.start(...args))
    //                      ^ this runs async
  }
}
```

To ensure the state update (the ripple being updated) runs inside `act()`, we must refactor tests like this:

```typescript
// before
fireEvent.click(menuitem)

// after
await act(async () => {
  fireEvent.click(menuitem);
});
```

Before I go replace the 77 instances of this issue, I was wondering if someone has an idea how to avoid this? I find it strange that it only happens in jsdom, I would expect it to happen in any environment.

Reply:
`act` is still somewhat mysterious to me. But I usually tend to avoid it altogether by reasoning which assertions will need to have effects run, or will asynchronously update. Wrapping these assertion with a `waitFor` makes the test more understandable to me:

```diff
diff --git a/packages/mui-material/test/integration/MenuList.test.js b/packages/mui-material/test/integration/MenuList.test.js
index e8f2cc066e..bf3202e25e 100644
--- a/packages/mui-material/test/integration/MenuList.test.js
+++ b/packages/mui-material/test/integration/MenuList.test.js
@@ -7,6 +7,7 @@ import {
   fireEvent,
   screen,
   programmaticFocusTriggersFocusVisible,
+  waitFor,
 } from '@mui/internal-test-utils';
 import MenuList from '@mui/material/MenuList';
 import MenuItem from '@mui/material/MenuItem';
@@ -493,17 +494,13 @@ describe('<MenuList> integration', () => {
       );
 
       const menuitem = getByText('Arizona');
-      // user click
-      act(() => {
-        fireEvent.mouseDown(menuitem);
-        menuitem.focus();
-      });
 
-      await act(async () => {
-        fireEvent.click(menuitem);
-      });
+      fireEvent.mouseDown(menuitem);
+      menuitem.focus();
+      fireEvent.click(menuitem);
+
+      await waitFor(() => expect(menuitem).toHaveFocus());
 
-      expect(menuitem).toHaveFocus();
       if (programmaticFocusTriggersFocusVisible()) {
         expect(menuitem).to.have.class('focus-visible');
       } else {
```

Reply:
🤔 Another way is to add `act` around the `menu.focus()`.

```diff
diff --git a/packages/mui-material/test/integration/MenuList.test.js b/packages/mui-material/test/integration/MenuList.test.js
index e8f2cc066e..1135ca9ccb 100644
--- a/packages/mui-material/test/integration/MenuList.test.js
+++ b/packages/mui-material/test/integration/MenuList.test.js
@@ -494,14 +494,10 @@ describe('<MenuList> integration', () => {
 
       const menuitem = getByText('Arizona');
       // user click
-      act(() => {
-        fireEvent.mouseDown(menuitem);
-        menuitem.focus();
-      });
 
-      await act(async () => {
-        fireEvent.click(menuitem);
-      });
+      fireEvent.mouseDown(menuitem);
+      await act(() => menuitem.focus());
+      fireEvent.click(menuitem);
 
       expect(menuitem).toHaveFocus();
       if (programmaticFocusTriggersFocusVisible()) {
```

Source:
* https://plainenglish.io/blog/you-probably-dont-need-act-in-your-react-tests-2a0bcd2ad65c
* https://github.com/threepointone/react-act-examples/blob/master/sync.md

Reply:
For the remaining failing tests, the tests aren't about the ripple so `act` isn't even useful; the warning about using `act` is really the only thing that make those tests fail. `waitFor` would indirectly solve the problem, not because it's waiting for anything useful but only because awaiting for it gives enough time for the next microtask to run (where the ripple callback is), but something like `await Promise.resolve()` or `await new Promise(r => setTimeout(r, 0))` would also be enough.

IIUC, `fireEvent.focus(element)` only dispatches a focus event at the element, while `element.focus()` actually focuses the element (sets the `document.activeElement`, etc) *and* dispatches a focus event though the DOM's logic.

Reply:
> `act` is still somewhat mysterious to me

It's implemented [here](https://github.com/facebook/react/blob/main/packages/react/src/ReactAct.js#L32-L42) and it ties a lot into react internals to [make things sync](https://github.com/facebook/react/blob/c4b433f8cb31d6f73d4a800fcf11ed55c8689daf/packages/react-reconciler/src/ReactFiberWorkLoop.js#L2316-L2325).

Reply:
> waitFor would indirectly solve the problem, not because it's waiting for anything useful but only because awaiting for it gives enough time for the next microtask to run

That doesn't seem to be the complete story, it looks like for this test React needs to flush some updates as well, otherwise the following would work:
```tsx
// user click
act(() => {
  fireEvent.mouseDown(menuitem);
  menuitem.focus();
});

fireEvent.click(menuitem);

await Promise.resolve();
```

replacing `await Promise.resolve();` with `await act(() => {})` does work.

The main issue seem to be calling `menuitem.focus()` inside of `act` without awaiting it, add an await on this line and everything starts working:

```tsx
await act(() => {
  fireEvent.mouseDown(menuitem);
  menuitem.focus();
});
```

But since all `fireEvent` APIs internally use `act`, only `menuitem.focus()` actually need to be wrapped and awaited, all other `act` calls in this test aren't necessary. That's the second snippet I propose.

From the [docs](https://react.dev/reference/react/act#await-act-async-actfn): 
> We recommend using `act` with `await` and an `async` function. Although the sync version works in many cases, it doesn’t work in all cases and due to the way React schedules updates internally, it’s difficult to predict when you can use the sync version.
> We will deprecate and remove the sync version in the future.

This means we actually need to remove sync usage of `act` altogether in the repo. 🤔 I'm not sure the [eslint plugin](https://github.com/testing-library/eslint-plugin-testing-library) catches that

Reply:
From [the docs](https://testing-library.com/docs/user-event/intro/), it seems like we should always be using `userEvent` to drive user events. We also have this file in the codebase: https://github.com/mui/material-ui/blob/next/packages-internal/test-utils/src/userEvent.ts
If someone can give me some context as to why `userEvent.ts` exists on our codebase it would be nice. With the information I have I would delete the existing `userEvent.ts`, then [setup](https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent) RTL's `userEvent` to be present in our `render()`, then replace all failing cases with this new approach.

Reply:
Between core, base UI and X, I only see this being used in X. cc @michaldudak to confirm?

The new approach looks great

Reply:
If I recall correctly, we are not using @testing-library/user-event because Sebastian felt it was a wrong abstraction.

Reply:
> e.g. https://github.com/testing-library/user-event/issues/350.

I may be reading it wrong, but that thread (including Sebastian) seems to advocate for using `user-
event` over `fireEvent`. Except for specific use-cases, e.g. when explicitly testing implementation details.

There are many valid differing viewpoints on testing, but in mine one of the primary concerns to optimize for when writing tests is refactorability. i.e. How much of the implementation can be changed without having to touch a test? It's about creating that pit of success. Now you can easily write correct implementation and break a bunch of tests, while the opposite should be true, most of the time when tests breaks it should mean that an incorrect implementation is supplied. That is when they effectively prevent maintainers from introducing regressions.

Reply:
I've started doing the changes for tests failing in jsdom, but I realize that `userEvent` won't work for many cases because many of them are unit-testing specific event handlers rather than the full interaction, so they need to keep using `fireEvent`. I could refactor our `fireEvent` wrapper to run `act(async () => ...)`, so the failing tests in jsdom would only need an additional `await` in many cases rather than the full `await act(async () => ...)`. Does anyone see a downside to including `act` inside `fireEvent`?

Reply:
I didn't change `fireEvent`, making it async changes its type too much.

Reply:
The `await act(async () => {` pattern looks great. I have converted Jan's draft issue to an open one: https://github.com/mui/mui-public/issues/174. "We will deprecate and remove the sync version in the future" sold me.

Reply:
The problem with `await act(async () => {` is that it triggers an error from the Testing Library ESLint plugin when wrapping `fireEvent` (or other Testing Library helpers).

<img width="640" alt="image" src="https://github.com/user-attachments/assets/18e5c64e-3e7b-4e9f-8b07-58a6aaabe812">

I wonder if we could have another pattern that is "compliant" with the ESLint plugin.

Reply:
Is the pattern wrong or is the eslint plugin wrong?

Reply:
I don't know. Testing Lib helpers don't need to be wrapped in `act`, so I'd say the pattern is wrong on paper. I normally defer to `waitFor` for async operations, but I'm not familiar with the lazy ripple implementation and why it needs `await act(async () => {`.

Reply:
Following seems to work for me:

remove all `act` around `fireEvent`. Change the `act` around `button.focus()` to `await act(async () => {`

```diff
       if (typeof Touch !== 'undefined') {
         const touch = new Touch({ identifier: 0, target: button, clientX: 0, clientY: 0 });
 
-        await act(async () => fireEvent.touchStart(button, { touches: [touch] }));
+        fireEvent.touchStart(button, { touches: [touch] });
         expect(onTouchStart.callCount).to.equal(1);
 
-        await act(async () => fireEvent.touchEnd(button, { touches: [touch] }));
+        fireEvent.touchEnd(button, { touches: [touch] });
         expect(onTouchEnd.callCount).to.equal(1);
       }
 
       if (canFireDragEvents) {
-        await act(async () => fireEvent.dragEnd(button));
+        fireEvent.dragEnd(button);
         expect(onDragEnd.callCount).to.equal(1);
       }
 
-      await act(async () => fireEvent.mouseDown(button));
+      fireEvent.mouseDown(button);
       expect(onMouseDown.callCount).to.equal(1);
 
-      await act(async () => fireEvent.mouseUp(button));
+      fireEvent.mouseUp(button);
       expect(onMouseUp.callCount).to.equal(1);
 
-      await act(async () => fireEvent.contextMenu(button));
+      fireEvent.contextMenu(button);
       expect(onContextMenu.callCount).to.equal(1);
 
       await user.click(button);
       expect(onClick.callCount).to.equal(1);
 
-      act(() => {
+      await act(async () => {
         button.focus();
       });
       expect(onFocus.callCount).to.equal(1);
 
-      await act(async () => fireEvent.keyDown(button));
+      fireEvent.keyDown(button);
       expect(onKeyDown.callCount).to.equal(1);
 
-      await act(async () => fireEvent.keyUp(button));
+      fireEvent.keyUp(button);
       expect(onKeyUp.callCount).to.equal(1);
 
-      act(() => {
+      await act(async () => {
         button.blur();
       });
       expect(onBlur.callCount).to.equal(1);
 
-      await act(async () => fireEvent.mouseLeave(button));
+      fireEvent.mouseLeave(button);
       expect(onMouseLeave.callCount).to.equal(1);
     });
   });
```

This should be compliant with the eslint plugin.

Reply:
> I don't know. Testing Lib helpers don't need to be wrapped in `act`, so I'd say the pattern is wrong on paper. I normally defer to `waitFor` for async operations, but I'm not familiar with the lazy ripple implementation and why it needs `await act(async () => {`.

The top-level comment of this thread should explain why the wrapper is needed. The events dispatched by `fireEvent` trigger asynchronous React state updates (the lazy ripple being mounted), and React will log a warning that any state update must be wrapped in `act` if the state update happens asynchronously without the wrapper, because it's indeed not in an `act` context.

Reply:
Lovely. That looks better! I just opened a PR for these changes: https://github.com/mui/material-ui/pull/42942